crates/turborepo-lib/src/process/child.rs

```diff
@@ -609,31 +609,36 @@ impl Child {
         // across a channel
         let (byte_tx, mut byte_rx) = mpsc::channel(48);
         tokio::task::spawn_blocking(move || {
-            let mut buffer = Vec::new();
+            let mut buffer = [0; 1024];
+            let mut last_byte = None;
             loop {
-                match stdout_lines.read_until(b'\n', &mut buffer) {
-                    Ok(0) => break,
-                    Ok(_) => {
-                        if byte_tx.blocking_send(buffer.clone()).is_err() {
+                match stdout_lines.read(&mut buffer) {
+                    Ok(0) => {
+                        if !matches!(last_byte, Some(b'\n')) {
+                            // Ignore as this fails as we already are shutting down
+                            byte_tx.blocking_send(vec![b'\n']).ok();
+                        }
+                        break;
+                    }
+                    Ok(n) => {
+                        let mut bytes = Vec::with_capacity(n);
+                        bytes.extend_from_slice(&buffer[..n]);
+                        last_byte = bytes.last().copied();
+                        if byte_tx.blocking_send(bytes).is_err() {
```

Comment:
is this sending the last byte twice? once here in `byte_tx.blocking_send(bytes)` and once in the other part of the match? Or does `bytes.last().copied()` _remove_ that last byte from `bytes`?

Reply:
The last byte only gets sent via `byte_tx.blocking_send(bytes)`. The other branch arm is triggered when `read` returns 0 meaning the stdout has been closed. In that case we check if the output for the process ended with a newline and add one only if it didn't.

We need to add the newline as otherwise task output can look mangled since the next line we print would be appended to the output line with no trailing newline.